tools/real_time_graphing.py

```diff
@@ -0,0 +1,302 @@
+
+        self.check_time = self.start_time = time()
+
+        # Initializes figure for real_time_graph
+        plt.rcParams['toolbar'] = 'None'  # Disable matplot toolbar
+
+        self.fig = plt.figure(figsize=(8, 6))
+        self.fig.canvas.set_window_title('Real Time Graphing')
+
+        self.fig.subplots_adjust(hspace=1, wspace=0.75)  # Avoid subplot overlap
+
+        # Threading
+        self.sleep_time = kwargs['sleep_time'] if 'sleep_time' in kwargs.keys() else 1e-1
+
+        self.thread_stop = kwargs['thread_stop'] if 'thread_stop' in kwargs.keys() else threading.Event()
+
+        self.thread_queue = Queue()
+
+    def run(self):
+        self.parse_rtg_config()
+
+        self.ani = animation.FuncAnimation(self.fig, self.plot_data, blit=False, interval=20, repeat=False)
+
+        threads = {
+            'reader': threading.Thread(target=self.read_data, args=(self.thread_queue,)),
+            'processor': threading.Thread(target=self.process_data, args=(self.thread_queue,))
+        }
+
+        for thread in threads.values():
+            thread.start()
+
+        plt.show()
+
+        #
+        # Cleanup
+        self.thread_stop.set()
+
+        for thread in threads.values():
+            thread.join()
+
+    @property
+    def config_filename(self):
+        """
+        Returns the filename of config file.
+        """
+
+        config_path = __file__.split(RealTimeGraph.working_filename)[0]  # Get this files location
+
+        config_path += RealTimeGraph.relative_config_path
+
+        return config_path
+
+    def parse_rtg_config(self):
+        """
+        Interprets the graph config file
+
+        Returns
+        -------
+        dict
+            Parsed config file
+        """
+
+        try:
+            output = parse_config(self.config_filename)
+        except IOError:
+            logging.warning("RTG: Failed to read config file!".format(self.config_filename))
+            output = None
+
+        if output is None:
+            logging.critical("RTG: No configuration file found!")
+            return
+
+        # Total number of subplots
+        graph_count = [graph["output"] == 'text' for graph in output].count(False)
+
+        nrows = max(min(graph_count, RealTimeGraph.max_rows), 1)
+
+        ncols = int(graph_count / nrows) + (graph_count % nrows)
+
+        def unique_color_generator(colors_taken):
+            color_list = ['blue', 'orange', 'red', 'green', 'yellow', 'black', 'Ran out of colors!']
+
+            seen = set(colors_taken)
+
+            for elem in color_list:
+                if elem not in seen:
+                    yield elem
+                    seen.add(elem)
+
+        for graph in output:
+            if graph['output'] == 'text':
+                i = 0
+                for metric in graph['metrics']:
+                    # Coords are percent
+                    text = ax.text(i * (1 / len(graph['metrics'])) + .01, .01, 'matplotlib', transform=plt.gcf().transFigure)
```

Comment:
.01 is unexplained here, probably should make it a constant. also maybe add a comment explaining what's going on